# 01 - Data Understanding

## Projecto: Supply Chain Control Tower

Antes de comenzar el análisis, es importante entender qué información contiene el dataset y si realmente nos permitirá responder a las preguntas de negocio planteadas.

En esta fase se revisará la estructura de los datos, la calidad de los datos y las principales variables relacionadas con inventario, proveedores, demanda y aprovisionamiento.

In [1]:
#1. Importar librerias
import pandas as pd

In [3]:
#2. Cargar dataset
df = pd.read_csv("../Datasets/supply_chain_dataset1.csv")

In [4]:
#3. Exploración dataset
df.head(10)

,Date,SKU_ID,Warehouse_ID,Supplier_ID,Region,Units_Sold,Inventory_Level,Supplier_Lead_Time_Days,Reorder_Point,Order_Quantity,Unit_Cost,Unit_Price,Promotion_Flag,Stockout_Flag,Demand_Forecast
0,2024-01-01,SKU_1,WH_1,SUP_8,West,10,592,14,379,0,13.95,20.48,0,0,8.52
1,2024-01-02,SKU_1,WH_1,SUP_8,West,17,575,14,379,0,13.95,20.48,0,0,18.63
2,2024-01-03,SKU_1,WH_1,SUP_8,North,35,540,14,379,0,13.95,20.48,1,0,39.62
3,2024-01-04,SKU_1,WH_1,SUP_8,South,24,516,14,379,0,13.95,20.48,0,0,19.43
4,2024-01-05,SKU_1,WH_1,SUP_8,West,21,495,14,379,0,13.95,20.48,0,0,18.70
5,2024-01-06,SKU_1,WH_1,SUP_8,East,18,477,14,379,0,13.95,20.48,0,0,18.33
6,2024-01-07,SKU_1,WH_1,SUP_8,North,19,458,14,379,0,13.95,20.48,0,0,18.91
7,2024-01-08,SKU_1,WH_1,SUP_8,North,23,435,14,379,0,13.95,20.48,1,0,25.84
8,2024-01-09,SKU_1,WH_1,SUP_8,West,25,410,14,379,0,13.95,20.48,0,0,21.70
9,2024-01-10,SKU_1,WH_1,SUP_8,South,25,385,14,379,0,13.95,20.48,0,0,21.34


In [5]:
df.shape

(91250, 15)

El dataset contiene 91.250 registros y 15 variables.
Se trata de un volumen de información suficientemente amplio para el análisis.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 91250 entries, 0 to 91249
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Date                     91250 non-null  str    
 1   SKU_ID                   91250 non-null  str    
 2   Warehouse_ID             91250 non-null  str    
 3   Supplier_ID              91250 non-null  str    
 4   Region                   91250 non-null  str    
 5   Units_Sold               91250 non-null  int64  
 6   Inventory_Level          91250 non-null  int64  
 7   Supplier_Lead_Time_Days  91250 non-null  int64  
 8   Reorder_Point            91250 non-null  int64  
 9   Order_Quantity           91250 non-null  int64  
 10  Unit_Cost                91250 non-null  float64
 11  Unit_Price               91250 non-null  float64
 12  Promotion_Flag           91250 non-null  int64  
 13  Stockout_Flag            91250 non-null  int64  
 14  Demand_Forecast          91250 no

No se observan valores nulos en ninguna de las columnas, lo que simplifica las fases de limpieza y transformación.

La única incidencia que vemos es el formato de la fecha, que deberá corregirse antes de realizar análisis temporales.

In [7]:
df.describe()

,Units_Sold,Inventory_Level,Supplier_Lead_Time_Days,Reorder_Point,Order_Quantity,Unit_Cost,Unit_Price,Promotion_Flag,Stockout_Flag,Demand_Forecast
count,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.0,91250.000000
mean,20.054564,471.522312,7.984000,300.068000,19.272493,12.203320,18.261800,0.101589,0.0,20.082033
std,9.068602,133.488002,3.907929,54.879945,82.340831,4.574982,7.121136,0.302109,0.0,9.503955
min,0.000000,168.000000,2.000000,201.000000,0.000000,5.020000,6.950000,0.000000,0.0,0.000000
25%,13.000000,370.000000,4.000000,252.000000,0.000000,8.180000,12.000000,0.000000,0.0,12.950000
50%,20.000000,461.000000,8.000000,300.000000,0.000000,11.990000,18.180000,0.000000,0.0,19.950000
75%,27.000000,564.000000,11.000000,346.000000,0.000000,16.320000,23.390000,0.000000,0.0,26.930000
max,59.000000,990.000000,14.000000,398.000000,499.000000,19.760000,35.100000,1.000000,0.0,61.420000


El nivel medio de inventario es de aproximadamente 472 unidades, mientras que las ventas medias son de 20 unidades por registro.

Los lead times de los proveedores oscilan entre 2 y 14 días.

In [8]:
#4. Calidad de los datos
df.isna().sum()

Date                       0
SKU_ID                     0
Warehouse_ID               0
Supplier_ID                0
Region                     0
Units_Sold                 0
Inventory_Level            0
Supplier_Lead_Time_Days    0
Reorder_Point              0
Order_Quantity             0
Unit_Cost                  0
Unit_Price                 0
Promotion_Flag             0
Stockout_Flag              0
Demand_Forecast            0
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(0)

No se han encontrado valores nulos ni registros duplicados.

In [10]:
#5. Variables claves
df["SKU_ID"].nunique()
df["Supplier_ID"].nunique()
df["Warehouse_ID"].nunique()
df["Region"].nunique()

df["SKU_ID"].nunique(), df["Supplier_ID"].nunique(), df["Warehouse_ID"].nunique(), df["Region"].nunique()

(50, 10, 5, 4)

El dataset contiene:

- 50 productos (SKU)
- 10 proveedores
- 5 almacenes
- 4 regiones

Esta diversidad permite realizar análisis comparativos entre productos, proveedores y almacenes.

In [11]:
#6. Fechas
df["Date"] = pd.to_datetime(df["Date"])
df["Date"].min()
df["Date"].max()

df["Date"].min(), df["Date"].max()

(Timestamp('2024-01-01 00:00:00'), Timestamp('2024-12-30 00:00:00'))

Los datos cubren prácticamente todo el año 2024, desde el 1 de enero hasta el 30 de diciembre.

Esto permitirá analizar tendencias mensuales, estacionalidad y evolución del inventario a lo largo del tiempo en Power BI. 

In [12]:
df["Stockout_Flag"].value_counts()

Stockout_Flag
0    91250
Name: count, dtype: int64

In [13]:
df[df["Order_Quantity"] > 0].shape

(5027, 15)

No existen registros de rotura de stock (Stockout_Flag = 1), por lo que este indicador no podrá utilizarse como KPI principal.

Sin embargo, sí existen 5.027 registros de reposición de inventario, lo que permite analizar el comportamiento de compras y aprovisionamiento.

## Conclusiones iniciales

Tras la exploración inicial del dataset se observa que:

- Los datos presentan una excelente calidad, sin nulos ni duplicados.
- El dataset contiene información suficiente para analizar inventario, proveedores y aprovisionamiento.
- Existen 50 productos, 10 proveedores y 5 almacenes.
- El periodo analizado cubre prácticamente todo el año 2024.
- No existen roturas de stock registradas.
- Sí existen eventos de reposición de inventario.
- El dataset es adecuado para desarrollar una aplicación de negocio orientada a la optimización de inventario y procurement.
- En fases posteriores se incorporará una API de tipos de cambio para analizar el impacto de la volatilidad de divisas en los costes de aprovisionamiento.

## Business Validation Analysis
El objetivo de esta sección es comprobar si los datos permiten responder a las preguntas de negocio definidas para el proyecto.

En concreto, se analizarán diferencias entre regiones, almacenes y proveedores para validar la viabilidad del proyecto de optimización de inventario y procurement.

In [15]:
#Ventas por region
df.groupby("Region")["Units_Sold"].sum()

Region
East     459961
North    459849
South    456954
West     453215
Name: Units_Sold, dtype: int64

Se observan diferencias en el volumen de ventas entre regiones, lo que permitirá realizar análisis comparativos y evaluar el comportamiento de la demanda en distintas zonas geográficas.

In [16]:
#Inventario medio por almacen
df.groupby("Warehouse_ID")["Inventory_Level"].mean().sort_values(ascending=False)

Warehouse_ID
WH_2    491.489644
WH_4    477.390301
WH_5    467.116329
WH_1    466.056219
WH_3    455.559068
Name: Inventory_Level, dtype: float64

El nivel medio de inventario varía entre almacenes, lo que sugiere posibles diferencias en la gestión del stock y en la estrategia de aprovisionamiento.

In [17]:
#Lead time medio por proveedor
df.groupby("Supplier_ID")["Supplier_Lead_Time_Days"].mean().sort_values(ascending=False)

Supplier_ID
SUP_4     8.640000
SUP_8     8.500000
SUP_2     8.423077
SUP_3     8.260870
SUP_9     8.187500
SUP_1     8.107143
SUP_6     7.900000
SUP_7     7.558824
SUP_10    7.407407
SUP_5     6.956522
Name: Supplier_Lead_Time_Days, dtype: float64

Los proveedores presentan diferentes tiempos de suministro, lo que puede afectar a las decisiones de reposición y al riesgo asociado al inventario.

In [18]:
#Crear Procurement Spend
df["Procurement_Spend"] = (
    df["Order_Quantity"] *
    df["Unit_Cost"]
)

In [19]:
#Gasto por proveedor
df.groupby("Supplier_ID")["Procurement_Spend"].sum().sort_values(ascending=False)

Supplier_ID
SUP_7     2786515.46
SUP_8     2783632.55
SUP_10    2377444.98
SUP_4     2344434.42
SUP_1     2316775.61
SUP_2     2105200.35
SUP_3     1926682.69
SUP_6     1813833.07
SUP_5     1781871.10
SUP_9     1231182.10
Name: Procurement_Spend, dtype: float64

El gasto en compras no se distribuye de forma uniforme entre proveedores. Esto permitirá identificar proveedores estratégicos y analizar posibles dependencias.

In [20]:
#Error de forecast 
df["Forecast_Error"] = (
    df["Units_Sold"] -
    df["Demand_Forecast"]
)

In [21]:
df["Forecast_Error"].describe()

count    91250.000000
mean        -0.027469
std          2.987263
min        -14.080000
25%         -2.030000
50%         -0.010000
75%          1.990000
max         13.000000
Name: Forecast_Error, dtype: float64

El análisis del error de forecast permitirá evaluar la precisión de la previsión de demanda y detectar posibles desviaciones entre la demanda estimada y la demanda real.

## ¿Tenemos suficientes diferencias entre regiones, proveedores y almacenes para construir una aplicación de negocio interesante?
- Las ventas se distribuyen de forma bastante homogénea entre regiones, por lo que esta variable no parece ser un factor diferenciador relevante para el análisis.
- Los almacenes presentan diferencias entre sí, pero no son enormes. Podemos utilizarlo como filtro o análisis secundario.
- En la parte de proveedores, sí hay valor, ya que la diferencia entre proveedores es relevante.
- Procurement Spend - La diferencia es enorme, por lo que nos podemos preguntar: 
1. Quien concentra el gasto?
2. Existe dependencia de proveedores?
3. Que proveedor es más estratégico?
- Forecast Error es prácticamente cero, por lo que podemos analizar productos con mayor error, proveedores asociados, almacenes asociados. 

## ¿Cómo quedaría la Control Tower en Power BI?
1. Executive Dashboard: revenue, inventory value, procurement spend, average lead time. 
2. Inventory Optimization: Inventory coverage, reporder point analysis, inventory value, replenishment frecuency.
3. Supplier & Procurement Analysis: Spend by Supplier, Lead time by Supplier, Supplier Ranking. 
4. Forecast Accuracy: Forecast Vs Actual, Forecast Error, Top SKUs con mayor desviación. 
5. FX Impact Analysis: tipos de cambio. 